ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ACSReader.scala

```diff
@@ -107,32 +119,97 @@ class FilterTableACSReader(
+      // Approximation how many index entries can be present in one btree index leaf page (fetching smaller than this only adds round-trip overhead, without boiling down to smaller disk read)
+      // Experiments show party_id, template_id index has 244 tuples per page, wildcard party_id index has 254 per page (with default fill ratio for BTREE Index)
+      // Picking a lower number is for accommodating pruning, deletions, index bloat effect, which boil down to lower tuple per page ratio.
+      val RecommendedMinIdPageSize = 200
+      val calculatedMaxIdPageSize =
+        idPageWorkingMemoryBytes
+          ./(8) // IDs stored in 8 bytes
+          ./(
+            idPageBufferSize + 1
+          ) // for each filter we need one page fetched for merge sorting, and additional pages might reside in the buffer
+          ./(filterSize)
+      if (calculatedMaxIdPageSize < LowestIdPageSize) {
+        logger.warn(
+          s"Calculated maximum ID page size supporting API stream memory limits [$calculatedMaxIdPageSize] is too low: $LowestIdPageSize is used instead. Warning: API stream memory limits not respected. Warning: Dangerously low maximum ID page size can cause poor streaming performance. Filter size [$filterSize] too large?"
+        )
+        IdQueryConfiguration(LowestIdPageSize, LowestIdPageSize)
+      } else if (calculatedMaxIdPageSize < RecommendedMinIdPageSize) {
+        logger.warn(
+          s"Calculated maximum ID page size supporting API stream memory limits [$calculatedMaxIdPageSize] is very low. Warning: Low maximum ID page size can cause poor streaming performance. Filter size [$filterSize] too large?"
+        )
+        IdQueryConfiguration(calculatedMaxIdPageSize, calculatedMaxIdPageSize)
+      } else if (calculatedMaxIdPageSize < maxIdPageSize) {
+        logger.info(
+          s"Calculated maximum ID page size supporting API stream memory limits [$calculatedMaxIdPageSize] is low. Warning: Low maximum ID page size can cause poor streaming performance. Filter size [$filterSize] too large?"
+        )
+        IdQueryConfiguration(RecommendedMinIdPageSize, calculatedMaxIdPageSize)
+      } else {
+        logger.debug(
+          s"Calculated maximum ID page size supporting API stream memory limits [$calculatedMaxIdPageSize] is sufficiently high, using [$maxIdPageSize] instead."
+        )
+        IdQueryConfiguration(RecommendedMinIdPageSize, maxIdPageSize)
+      }
+    }
+  }
+
   def idSource(
-      pageBufferSize: Int
-  )(getPage: Long => Future[Vector[Long]]): Source[Long, NotUsed] = {
+      idQueryConfiguration: IdQueryConfiguration,
+      pageBufferSize: Int,
+  )(getPage: IdQuery => Future[Array[Long]]): Source[Long, NotUsed] = {
     assert(pageBufferSize > 0)
     Source
-      .unfoldAsync(0L) { fromExclusive =>
-        getPage(fromExclusive).map {
+      .unfoldAsync(
+        IdQuery(0L, Math.min(idQueryConfiguration.minPageSize, idQueryConfiguration.maxPageSize))
```

Comment:
When would a min be bigger than max? Should this rather assert?

Reply:
Good point, initially I thought we don't care, but you are right: constraining the IdQueryConfiguration gives us simpler logic here and one less unit test case